### DISCUSSION TITLE
Obtaining the type under the cursor?

### DISCUSSION THREAD

**Discussion by digitaldias** [VsixCommunity/Community.VisualStudio.Toolkit#400] [Q&A] [answered]
I am trying to write a command that mimics what Rider does in `Navigate to consumed API`, i.e. if I have a command named `QueryPeople`  that I can invoke using `Mediator.Send`, then, when I'm looking at code like this: 

```csharp
var query = new QueryForPerson(name: "Gringo");
var result = await _mediator.Send(query, cancellationToken);
```
What I'd like my extension to do is when I have the cursor on the `query` on line two, I press my shortcut, and I can then somehow retrieve the type of `query`, and then look up any C# classes that have a `Handle` method for that type, in this case, a `QueryForPersonHandler` which then has the appropriate method: `public async Task<something> Handle(QueryForPerson query, CancellationToken ct)`

Thsi is my first time writing an extension for VS2022, I've watched @madskristensen s video introduction, and fell short immediately as the `ActiveDocument` does not appear to give me what I need. I'll do some google-fu meanwhile, but if you know how I can obtain this type, or even how I should approach this problem and can give me a few pseudo-code suggestions, I'd really appreciate it.

**Comment by reduckted** [comment, marked as answer]
First, you'll need to install the [`Microsoft.VisualStudio.LanguageServices`](https://www.nuget.org/packages/Microsoft.VisualStudio.LanguageServices/) NuGet package.

Then you can get the document in Roslyn like this:

```cs
// Get the workspace.
Workspace workspace = await VS.GetMefServiceAsync<VisualStudioWorkspace>();

// Get the active Visual Studio document.
DocumentView? documentView = await VS.Documents.GetActiveDocumentViewAsync();

if (document is not null) {
    // Get the Roslyn document for that file.
    DocumentId? id = workspace.CurrentSolution.GetDocumentIdsWithFilePath(documentView.FilePath).FirstOrDefault();

    if (id is not null) {
        Document roslynDocument = workspace.CurrentSolution.GetDocument(id);
    }
}
```

If you know your way around Roslyn, then it's probably not too difficult to get the type at the cursor. I haven't tried this, but it might point you in the right direction.
```cs
// Get the selection.
int position = documentView.TextView.Selection.ActivePoint.Position.Position;

// Get the syntax root.
SyntaxNode root = await roslynDocument.GetSyntaxRootAsync();

// Find the node at the selection.
SyntaxNode selectedNode = root.FindNode(new TextSpan(position, 1));

// Get the type information for that node.
SemanticModel model = await roslynDocument.GetSemanticModelAsync();
TypeInfo typeInfo = model.GetTypeInfo(node);
```

**Reply by digitaldias** [reply]
Thank you so much! It definetly got me closer, the typeInfo does not provide me with a type, but I see that the node is the correct one, so I can do some more research from there. Kudos for a very detailed reply!
